### ISSUE TITLE
Can't compile js file. ReferenceError: identifier 'window' undefined

### ISSUE THREAD

**Issue by AngelTsTsvetkov** [svaarala/duktape#1530]
Error In attached file on 72line:
var Crypto = window.Crypto = {};

Error:
ReferenceError: identifier 'window' undefined

[test.zip](https://github.com/svaarala/duktape/files/1018186/test.zip)

**Comment by svaarala** [comment]
`window` is not part of the Ecmascript standard bindings and is browser specific. You can work around it with:

```js
Object.defineProperty(new Function('return this')(), 'window', {
    value: new Function('return this')(),
    writable: false, enumerable: true, configurable: false
});
```

The new Function() calls are a way to get the global object reference reliably regardless of strictness etc.

**Comment by AngelTsTsvetkov** [comment]
Thank you so much, this resolved that error, but unfortunately I thinks this file can't be start with duktape. There are so much errors. One i resolve but stuck to next one. Can you help?
TypeError: undefined not callable
    at [anon] (duk_js_call.c:775) internal
    at ECB (test.js:5865) construct preventsyield
    at buildVersions (test.js:6051)
    at global (test.js:5991) preventsyield
error in executing file test.js

**Comment by fatcerberus** [comment]
The stack trace indicates the error is with a `__defineGetter__` call.  Duktape doesn't yet support that function, but there's an issue open to add support: #1529

**Comment by svaarala** [comment]
@AngelTsTsvetkov Duktape tries to provide a plain Ecmascript environment and most browser/legacy bindings are not directly supported. Unfortunately some JS code is more browser oriented and that can be a limiting issue; you'd then need to fill in the missing bindings.

There are some "polyfills" in the Duktape repo, here https://github.com/svaarala/duktape/tree/master/polyfills. For example, `__defineGetter__` and `__defineSetter__`:
- https://github.com/svaarala/duktape/blob/master/polyfills/object-prototype-definegetter.js
- https://github.com/svaarala/duktape/blob/master/polyfills/object-prototype-definesetter.js

**Comment by AngelTsTsvetkov** [comment]
svaarata, thank you so much for your help. Because i see you are tp professional in js engine field, could you be so kind to recommend some js engine, which can run my attached above file test.js? This is WalletGenerator.net js. I tired V7 but error rise (it is not syntax). V8 or Nodejs or tomcat. I can't use because of sharing hosting linux plan i am using. Thank you so much for your time and support.

**Comment by svaarala** [comment]
@AngelTsTsvetkov Just quickly looking at the code, it seems to be browser Javascript (it finds elements, works on their .innerHTML) so it would only work in a browser. If those sections of the code are not relevant, it might be runnable by stripping away the browser specific parts. I don't think you'd have much more success with any "pure Ecmascript" engine like Duktape (AFAICT v7 is in the same category) because such engines don't provide browser bindings.

**Comment by fatcerberus** [comment]
The code in question appears to be a polyfill of some sort.  In that case making it work would just be a matter of removing the browser-specific polyfills (since they're not relevant outside of a browser anyway).

Alternatively, one could use core-js if ES2015+ builtins are what is needed.

**Comment by AngelTsTsvetkov** [comment]
I don't need browser-specific function. and removed most part of it. this code generate bitcoin (130+cryptocurrencies) public and private address.

**Comment by svaarala** [comment]
@AngelTsTsvetkov Let us know how it turned out, happy hacking!

**Comment by AngelTsTsvetkov** [comment]
not well actually :(. With duke, after remove browser-specific code, i didn't success. I don't know how to use "core.js". So I am stuck :(.

**Comment by svaarala** [comment]
What are the binding(s) still missing?

**Comment by AngelTsTsvetkov** [comment]
Guys, I did it.
When I managed to remove absolutely all browser-specific codes (like document.*, window.*, etc.) and set 3 tricks respectively for "window", "__defineGetter__" and "__defineSetter__". the code is executed without a problem and like a charm :).
Very big and special thanks to svaarala and fatcerberus.- without your help I will never, ever resolve that task.
duktape-2.1.0 work perfect, fast and clean and I love it. Keep extend it!
Thank you again and God bless!

**Comment by svaarala** [comment]
@AngelTsTsvetkov Awesome, great to hear -- `__defineGetter__` and `__defineSetter` will probably be included in 2.2 release so you'd then only need `window` :)

**Comment by AngelTsTsvetkov** [comment]
svaarala, can you tell me how to input a variable from shell into my script?Is it possible with duktape2.1.0?
./duk test.js -someVariable

**Comment by svaarala** [comment]
You can eval snippets before the file, e.g.:

```js
// test.js
print(FOO);
```

and execute as:

```
$ duk -e 'var FOO=123' /tmp/test.js
123
```

You can also set up the variable from C code -- in general, C code can operate on the global object, other objects, their properties, etc, just like JS code. For example, to set up FOO=123:

```c
duk_push_int(ctx, 123);
duk_put_global_string(ctx, "FOO");

/* Then evaluate test.js. */
```

**Comment by AngelTsTsvetkov** [comment]
Unfortunately I can't reproduce it. Here is an example:

test.js:
var DefaultCoin="Bitcoin";

action.c:
...
push_file_as_string(ctx, argv[1]);
duk_safe_call(ctx, eval_raw, NULL, 1 /*nargs*/, 1 /*nrets*/);
duk_safe_call(ctx, tostring_raw, NULL, 1 /*nargs*/, 1 /*nrets*/);
res = duk_get_string(ctx, -1);
duk_pop(ctx);

$./duk -e 'var DefaultCoin=Litecoin' test.js

______________________________________________
the result is always Bitcoin, not Litecoin. So never changed?

**Comment by svaarala** [comment]
I'm not sure how the script works but if test.js includes `var DefaultCoin = 'BitCoin'` *and* then proceeds to run directly inside the same eval() call, you don't have a place where you could overwrite the default: if you set it before and the script says `var DefaultCoin = 'BitCoin'` it gets overwritten. Doing the assignment afterwards obviously won't have an effect.

If this is the case:
- You can change the script so that it checks if DefaultCoin is set, and only assigns 'BitCoin' if not, e.g. `var DefaultCoin = DefaultCoin || 'BitCoin';`. Then if you say `var DefaultCoin='LiteCoin'` before the script, it will take effect for the rest of the script.
- Or you can change the script so that when you eval() it, it just registers the variables and functions but doesn't yet actually run anything. Once that initial eval() returns you'll then be able to modify the global variables like `DefaultCoin`. Finally, you'd invoke some function within the script to actually run the functionality, say `run()`.

But I'm not sure what the intended behavior is.

**Comment by AngelTsTsvetkov** [comment]
svaarala, once again - thank you so so much for your detailed info and support. You was right. My mistake was that I set 'var DefaultCoin = "Bitcoin";' in the beggining of js file. I remove it and then I was able to change that js variable within your two ways:
1.as argument:
$ ./action 'Litecoin' test.js
and in action.c:
...
duk_push_string(ctx, argv[1]);
duk_put_global_string(ctx, "DefaultCoin");
...

2.as C variable:
$ ./action test.js
and in action.c:
...
char *h="MagicInternetMoney";
duk_push_string(ctx, h);
duk_put_global_string(ctx, "DefaultCoin");
...


I really appreciate your help. I will be glad If I can help you by any ways. Regards
